line chart: update only visible charts

stephanwlee opened the pull request:
In the timeseries dashboard, we keep track of cards that are in the
view. With this information, we can opt out of updating the charts that
are outside of the viewport and save CPU cycles.

This change propagates the visibility information via `enableUpdate`
flag and make sure to apply changes only when the card is visible in the
viewport.

psybuzz commented on tensorboard/webapp/widgets/line_chart_v2/line_chart_component.ts:
```diff
@@ -107,6 +107,14 @@ export class LineChartComponent implements AfterViewInit, OnChanges, OnDestroy {
+   * changed and applies the change when the update is enabled.
+   */
+  @Input()
+  enableUpdate?: boolean;
```
Could we invert the meaning and have a property `disableUpdates` instead?  My default assumption when reading optional properties is that `enableXYZ = false` would be equivalent to not specifying a value.

Then L319 can be `if (!this.lineChart || this.disableUpdate) return;`

stephanwlee replied:
I generally prefer positive APIs but I can see a case of `disableUpdate` here, too.

I do not exactly recall but I think there was ToTT or code style guide around this. It says something like how `disableUpdate=false` making you think more.

psybuzz commented on tensorboard/webapp/widgets/line_chart_v2/line_chart_component.ts:
```diff
@@ -307,7 +316,7 @@ export class LineChartComponent implements AfterViewInit, OnChanges, OnDestroy {
    */
   private updateLineChart() {
-    if (!this.lineChart) return;
+    if (!this.lineChart || this.disableUpdate === true) return;
```
nit: Can we do a simpler `if (!this.lineChart || this.disableUpdate) return;`?

psybuzz commented on tensorboard/webapp/widgets/line_chart_v2/lib/drawable.ts:
```diff
@@ -23,15 +23,26 @@ import {
+export interface RenderCachePublicApi {
+  getFromPreviousFrame(key: string): Cacheable | null;
+  setToCurrentFrame(key: string, value: Cacheable): void;
+}
```
Would it be cleaner to move this interface definition into `paint_brush.ts` or another common file, to avoid a circular dependency between `drawable.ts` and `paint_brush.ts`?

psybuzz commented on tensorboard/webapp/widgets/line_chart_v2/lib/paint_brush.ts:
```diff
@@ -40,7 +38,9 @@ export class PaintBrush {
       paintOpt
     );
-    this.context.renderCache.setToCurrentFrame(cacheId, newCacheValue);
+    if (newCacheValue) {
```
Looking at paint_brush.ts's signatures alone, I'd expect that `setLine(/* args for valid line */)` followed by `setLine(/* args for null line */)` to result in null or no entry in the current frame map.  With this new condition, we are saying that if `setLine` fails to create a valid line, it will preserve the existing one.

Could you help me understand when why this change is needed?

stephanwlee commented:
Sorry about the confusion. I mistakenly added commit that appeared in a different PR that is already merged. Rebased and omitted and removed unnecessary noise.

psybuzz reviewed (approved):
Looks good!
